bundles/org.openhab.binding.hue/src/main/resources/OH-INF/thing/Clip2Thing.xml

```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="hue" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- API v2 device thing -->
+	<thing-type id="device">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bridge-api2"/>
+		</supported-bridge-type-refs>
+
+		<label>Hue Device</label>
+		<description>A Hue API v2 device with channels depending on its actual capabilities.</description>
+
+		<channels>
+			<channel id="brightness" typeId="light-brightness"/>
+			<channel id="color" typeId="system.color"/>
+			<channel id="color-temperature" typeId="system.color-temperature"/>
+			<channel id="color-temp-kelvin" typeId="system.color-temperature-abs"/>
+			<channel id="switch" typeId="light-switch"/>
```

Comment:
Yes, we do absolutely need these channels. This is probably the most fundamental difference between v1 and v2. And is therefore probably the fundamental reason why Philips (Signify or whatever) introduced the v2 API.

In v1 there are seven different hard coded bulb types resp. thing types where each different thing type has a different functionality according to the [TABLE](https://www.openhab.org/addons/bindings/hue/#supported-things). Philips / Signify realised that such a hard coded functionality is impossible to maintain going forwards.

Therefore in v2 each type of light may or may not support different functionality where the different options (color, switch, brightness, **and future options** etc.) are defined in API DTOs, but the combination of those options in a given light or group of lights, (now or in future), is NOT hard coded to anything, but rather is declared at runtime. 

So precisely for the above reason, the v2 model does NOT adopt the seven hard coded light type model of v1. Instead each light may (or may not) have a different combination of channels that are dynamically created at run time depending the actual self declared capabilities of the respective light.

Reply:
But I hope that for a color light device, you will keep only the color channels and not switch and brightness channels ?
In other words, switch and brightness channels will be used/kept only for non color lights ?

Reply:
> for a color light device, you will keep only the color channels and not switch and brightness channels

I think all channels MUST be present on such a device. There are a number of reasons..

Firstly v2 API has state entities for ColorXy, Dimming, OnState, ColorTemperature; whereas OH has DIFFERENT state entities HSBType, PercentType, OnOffType; and there is often not always an explicit 1:1 mapping between them. e.g. 
- If a Hue event message arrives with a new XY value, OH cannot automatically update the OH Color channel HSBType value, instead OH has to assemble the Color channel HSBType value from the new XY event value plus the PRIOR cached Dimming channel value.
- A Hue event message Dimming=0 does NOT always mean that the Switch channel OnOffType is Off; instead if Hue sends OnState=false then the Switch channel is explicitly OnOffType.Off

Secondly, in general I think applying artificial mappings between clearly not the same states, is dangerous, and could leads to potential future code spaghetti. It is safer to keep separate concepts separate.

Thirdly, I can imagine that if a user wants to (say) turn off all (types of) lights, it is easier to do the following rather than having to figure out which channel Id needs to be used for different types of lights..

```
for (light : lights) {
  sendCommand(light:switchChannel, OFF);
}
```
Finally, anyway, what harm does it make to keep all channels? You don't have to use them.

Reply:
> Finally, anyway, what harm does it make to keep all channels? You don't have to use them.

It "breaks" the original concept and duplicates channels rather having different items for one unique channel.
@kaikreuzer @jlaur @cweitkamp : please let us know if you think it is acceptable or not to have 3 channels color / brightness / switch to control a color light.

Reply:
I don't really see the reason for multiple channels either, given HSBType extends PercentType, and the fact that both color and dimmer channels are supposed to accept OnOffType.
I can see value in switch being separate from dimmer/color (to differentiate 'off' from 'on at 0 brightness', if that's actually possible? The hue app only allows going down to 1% brightness for me...), but I fail to see how a separate dimmer channel could be useful for a color light.

IMO, all 3 should be present in the thing type (because the capabilities are determined at runtime), but 2 of the 3 channels should be removed from the thing after determining the capabilities.

Reply:
> It "breaks" the original concept 

YES. This is a new API. It IS a BREAKINg change.

Reply:
>  differentiate 'off' from 'on at 0 brightness', if that's actually possible

Yes, that is indeed possible.

Reply:
@seime as you are a real user with a large system, can you please also give your inputs on this?

Reply:
Maybe also just to reiterate: even if OH would expose one single Color HSB channel, the binding would still have to internally maintain state for XY and Dimming (and possibly OnState too). SSE events arrive asynchronously at different times. There may first be an SSE event with (say) a Dimming value. And then some time later an SSE event for an XY value. And only if both SSE events have been combined, is it possible to reassemble the current Color HSB channel value. So the issue is not whether the binding internally **_maintains_** two channels (it MUST do so), but whether it **_exposes_** two channels.

Reply:
> have 3 channels color / brightness / switch to control a color light

And the issue is not just to **_control_** the three state values.

- I want to have a sitemap that simply shows if my lights are on or off. I do not want to eff around with mapping the B value from an HSB channel to a Switch that may of may not actually tell me if the light is on or off.
- Also I want to send a command to the light that explicitly turns it off, and not maybe sets is to a B value "somewhere close to zero"..

Reply:
As 'channel' is an OpenHAB concept, the binding doesn't need to maintain multiple channels, but just (Hue API) *values*. 
Just to make sure: when I talk about *channels*, I mean the (user visible) thing channels. Everything else is an implementation detail.

Reply:
You guys may have noticed that I feel strongly about this. I have been using this binding for six months, and I know what I am talking about. The only compromise that I would be prepared to make would be to have the channels that you don't like to see be marked as 'advanced'. But I am not willing to sacrifice them. I think that would be stupid.

Reply:
> Yes, that is indeed possible.

How though?

Reply:
> acceptable or not to have 3 channels color / brightness / switch to control a color light

Note: it is not that the user has to issue three separate commands -- of course one single command is sufficient. But the parts of the single command are split to act on the three channels..

You probably have not reviewed that part of the code, but you will see that if the user gives an HSB value as a command to the Color channel, the binding will convert the HS part of the command to an XY value, the B part to a Dimming value, and add an OnState=true. And when the command is received, the Hue will send one or more SSE events confirming the new XY, the new Dimming, and the new OnState. And the binding will assemble/dissemble those into the HSB, the Dimmer, and the Switch channels respectively.

Reply:
> > It "breaks" the original concept
> 
> YES. This is a new API. It IS a BREAKING change.

It is not what I mean. As @maniac103 said, you are breaking an openHAB concept (openHAB philosophy about channels).
The Hue binding will then become the only binding diverging from this OH concept for light control.
It could introduce some confusion to users.
That is the reason why it has to be seriously discussed. I am sure you were aware of this concept, so you should have discussed that at the beginning of your implementation.
I am not understand clearly your reasons but they may be valid. So we need advices from other maintainers.

Reply:
Look, it almost works already without the excess channels:
![image](https://user-images.githubusercontent.com/19519842/235765553-d8f5a2cc-277e-41e5-b31a-ad2f8427cba5.png)

When I use the dimmer to control, the other channels follow:
![image](https://user-images.githubusercontent.com/19519842/235765801-f194411e-faf7-436c-823e-742a1b01d42b.png)

I noticed though that commands for `OnOffType` does not work for this channel, so this must be fixed. Otherwise the linked `Switch` cannot be used to turn the light on or off. It's correctly reflected the other way.

When this is fixed: Why would there be a need for the Dimmer and Switch channels for a color light?

Reply:
You would need something like this in `Clip2ThingHandler`:

```java
            case HueBindingConstants.CHANNEL_2_COLOR:
                if (command instanceof OnOffType) {
                    putResource = new Resource(lightResourceType).setSwitch(command);
                } else if (command instanceof PercentType) {
                    putResource = new Resource(lightResourceType)
                            .setSwitch(PercentType.ZERO.equals(command) ? OnOffType.OFF : OnOffType.ON)
                            .setBrightness(command);

                } else {
                    putResource = new Resource(lightResourceType).setSwitch(OnOffType.ON).setColor(command,
                            getCachedResource(lightResourceType));
                }
                break;
```

Reply:
> I noticed though that commands for OnOffType are not handled for this channel, so this must be fixed

Note that Kai provided a fix yesterday: openhab/openhab-core#3588. Note sure if it is related but maybe.

Reply:
I don't think it's related to this scenario, Kai's case is more complex. With the code paste above added, it seems to work more or less.

Reply:
> Look, it almost works already without the excess channels

Yes. That is really my point. I will fix it so that the 'almost' is resolved. But nevertheless, I insist on retaining the (what you call 'excess') channels.

Reply:
> would need something like this in Clip2ThingHandler

Actually I think it should be in `Resource.java` but otherwise I agree with you.

Reply:
> What advantage does choice between 'send dimmer commands to dimmer channel' have over 'send dimmer commands to color channel' have?

- To the color channel you can send HSBType, PercentType, UpDownType, OnOffType
- To the dimmer channel you can send PercentType, UpDownType, OnOffType
- To the switch channel you can send OnOffType

All variants cause all affected channels to update.

The benefits are..
- If you just want to see on/off status you can see it without having to may HSB -> B -> OnOff (maybe)
- If you want to switch all lights off, you can `for (light : lights) { sendCommand(light:switchChannel, OFF) }`
- If you want to dim all lights, you can `for (light : lights) { sendCommand(light:dimmerChannel, 23%) }`
- When SSE sends partial updates e.g. on, dim, color, the respective channels will be updated directly (and when all parts are received, then all combinations (i.e. HSB) are reassembled.
- I want it so. :)

Reply:
^
So here are two proposals for hiding what you call ‘unwanted’ channels..

 ### Dynamic hiding

When a superior channel is dynamically created, the inferior channels dynamically become ‘advance’ channels and therefore not visible to normal users. I.e. if Color channel is created, then on/off and brightness become advanced and thus hidden. Advantage is no user action required.

### Static hiding

Set a config param on the binding ‘idiot mode / advanced mode’ in which the ‘unwanted’ channels are not created in the former mode. Disadvantage needs more explanation.

Reply:
Come back to the basics ;)
A channel in OH can support several kinds of commands. For example, a color channel should support of course a color value but also a dim value (percentage) and an ON/OFF value.
Now come back to both of your arguments I do not understand:

> If you want to switch all lights off, you can for (light : lights) { sendCommand(light:switchChannel, OFF) }
> If you want to dim all lights, you can for (light : lights) { sendCommand(light:dimmerChannel, 23%) }

What is expected is that you can do it like that:
for (light : lights) { sendCommand(light:colorChannel, OFF) }
for (light : lights) { sendCommand(light:colorChannel, 23%) }

> It makes me happy :)

That is a serious argument but not sure it is sufficient :)

Reply:
And regarding your other benefits:

> If you just want to see on/off status you can see it without having to map HSB -> B -> OnOff (maybe)
> If you just want to see brightness status you can see it without having to map HSB -> B (maybe)

What is exactly the problem with that ?
I believe it is done like that in all light bindings.
Or maybe I do not understand what you mean.

> When SSE sends partial updates e.g. on, dim, color, the respective channels will be updated directly (and when all parts are received, then all combinations (i.e. HSB) are reassembled.

Looks like a technical "problem". Of course, SSE events (I assume you mean events coming from the hue bridge) have to be properly handled to update all channels.

Reply:
> User is not forced to make a choice..
> 
>     * To the color channel you can send HSBType, PercentType, UpDownType, OnOffType
> 
>     * To the dimmer channel you can send PercentType, UpDownType, OnOffType
> 
>     * To the switch channel you can send OnOffType

So you know well the concept :)
Having only one channel to control its light is more simple for a user.
Then he can link it to different kinds of items if he wants a switch or a dimmer (brightness) control.
My feeling is that you are more or less mismatching channels and items.

Reply:
I have done following things..
- Changed the command logic as suggested by @jlaur above
- Changed SSE update logic to force B of HSB and Dimmer to 0 when SSE event Off received
- If Color channel is present then Dimmer and Switch are moved to Advanced category
- If Dimmer channel is present then Switch is moved to Advanced category

I believe this addresses all your concerns so I am now marking this thread as Resolved.

Reply:
I believe we still lack the answer to the question by @maniac103 - "Is it possible to create a group that includes both dimmer and color items? Whether the argument above holds or is rather weak probably depends on the answer to that question..."

And no matter the outcome of that, I believe we still didn't conclude the discussion for the initial comment by @lolodomo, i.e. whether there should be more than one channel according to openHAB conventions and concepts.

@J-N-K / @kaikreuzer - I vaguely remember at least one of you providing some insights for this specific question, but I can't seem to find it. Can you recall that?

Reply:
Look. Let me make this clear. I am the author of this code. And I want this functionality. If you will forbid that, then I shall be forced to withdraw this PR and withdraw my authorization for anyone else to use this code.

Reply:
You are apparently really blocked on that point.
Is it just because you want this like that even if it does not follow the OH standard ?
Or is it because there is a real constraint we have no yet understood ? In that case, sorry for that, please be patient and didactic and explain why the usual way to do is no more applicable for hue with the API v2.

We are just discussing between gentlemen, that the role of reviewers to check that OH standards are respected. So please keep cool and take the time to convince us why it is necessary to workaround the OH standard for this binding.

Reply:
In case it is only because you prefer your approach rather than the OH standard, to disarm the potentially blocking situation, I could propose the following solution: the binding should support the two modes and you can add a binding setting named for example "provideMoreChannels" with the following description "By enabling this setting, the binding will derogate from the openHAB standard and make available more channels than necessary to control the hue lights. For a color light thing for example, as usual a color channel will be available allowing to control the color of the light, its brightness and to turn it on/off, but in addition a dimmer channel and a switch channel will be also available allowing to adjust the brightness of the light and turn it on/off without using the color channel.". Of course, this option should be off by default, meaning the openHAB standard is respected. But then, you will have just to enable this option to switch to your preferred option to have redundant channels.

Reply:
Maybe we could have a Zoom call or something so we can talk about it?

Reply:
> you would need something like this in Clip2ThingHandler:

@jlaur actually it looks like this (lines 328 .. 364)

https://github.com/openhab/openhab-addons/blob/639d29d9f2063751367e63df92554e990d093cfb/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2ThingHandler.java#L328

And in the opposite direct the following..

https://github.com/openhab/openhab-addons/blob/639d29d9f2063751367e63df92554e990d093cfb/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java#L294

https://github.com/openhab/openhab-addons/blob/639d29d9f2063751367e63df92554e990d093cfb/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resource.java#L236

Reply:
> Firstly v2 API has state entities for ColorXy, Dimming, OnState, ColorTemperature; whereas OH has DIFFERENT state entities HSBType, PercentType, OnOffType; and there is often not always an explicit 1:1 mapping between them. e.g.

Can you please elaborate where there is an ambiguity or inconsistency how `OnOffType`/`PercentType`/`HSBType` are handled by a `Color` item/channel? The only thing we can't properly model with one item/channel is a light that supports color AND color-temperature. In that case we need two channels/items. 

> 
> * If a Hue event message arrives with a new XY value, OH cannot automatically update the OH Color channel HSBType value, instead OH has to assemble the Color channel HSBType value from the new XY event value plus the PRIOR cached Dimming channel value. And vice-versa for an incoming Dimming event value..
> * A Hue event message Dimming=0 does NOT always mean that the Switch channel OnOffType is Off; instead if Hue sends OnState=false then the Switch channel is explicitly OnOffType.Off

Maybe the misunderstanding here is that the item state exactly matches all properties of the device. I don't think that is the case. What is does it represents the state of the device. If brightness=0 it does not matter if the device internally has "power=true" or "power=false", the brightness is 0 and it's fine to represent that as "color=black". 

> Secondly, in general I think applying artificial mappings between clearly not the same states, is dangerous, and could lead to potential future code spaghetti. It is safer to keep separate concepts separate.

Where do you see the "artificial" mapping (I guess that is the same question as above)? In fact, I think that it is far more confusing to have a "switch" channel which is "off" and a "brightness" channel that is "50%". 
 
> Thirdly, I can imagine that if a user wants to (say) turn off all (types of) lights, it is easier to do the following rather than having to figure out which channel Id needs to be used for different types of lights..
> 
> ```
> for (light : lights) {
>   sendCommand(light:switchChannel, OFF);
> }
> ```

That is perfectly possible for a `color` channel. You can either link `Switch` items to that or just send on/off to a `Color`/`Dimmer` item. That is the whole idea here: `Dimmer` accepts everything that `Switch` would accept, `Color` accepts everything that `Dimmer` accepts. If you just want to use `Switch` functionality, that is perfectly possible.

If you separate e.g. brightness and power, it could be that a user drags the brightness slider to 70% and the lamp will not turn on (because you still have power=off). I believe that is unexpected for 99% of the users. 

What you can do (and what I do in some bindings, where the API supports it): If I receive an `OnOffType`, I just send that, a `PercentType` checks if the value is 0 (in that case it's off) and I send "off" and "0", if the value is larger than 0, I send "on" and the value. That ensures the reaction of the device is like the user expects.